src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -868,9 +869,38 @@ private void joinTokenRing(int delay) throws ConfigurationException
             setMode(Mode.WAITING_TO_BOOTSTRAP, "Awaiting start bootstrap call", true);
             bootstrapManager.awaitBootstrappable();
             if (SystemKeyspace.bootstrapInProgress())
-                logger.warn("Detected previous bootstrap failure; retrying");
+            {
+                // on detection of previous bootstrap failure prevent boostrap from proceeding if data directories are
+                // filled up beyond palantir_cassandra.bootstrap_disk_usage_threshold_percentage
+                Integer threshold = Integer.getInteger("palantir_cassandra.bootstrap_disk_usage_threshold_percentage");
```

Comment:
Nit: not a huge deal in this case, but our style for pulling env variables is to do it as a static field so that it only happens one time, and not in the critical path for the function.